### ISSUE TITLE
Out-of-memory issue during instantiation in IDEAS library

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#9009]
Some models of the IDEAS library, e.g. [IDEAS.Examples.PPD12.Heating](https://libraries.openmodelica.org/branches/master/IDEAS/files/IDEAS_IDEAS.Examples.PPD12.Heating.err) fail with

```
Warning: Out of memory! Faking a stack overflow.
Error: Stack overflow occurred while evaluating translateModel(IDEAS.Examples.PPD12.Heating, tolerance = 1e-06, outputFormat = "empty", numberOfIntervals = 5000, variableFilter = "", fileNamePrefix = "IDEAS_IDEAS.Examples.PPD12.Heating"):
[bt] #1 /var/lib/jenkins/ws/OpenModelicaLibraryTestingWork/OpenModelica/OMCompiler/build/bin/../lib/x86_64-linux-gnu/omc/libOpenModelicaRuntimeC.so(mmc_setStacktraceMessages_threadData)
[bt] #2 /var/lib/jenkins/ws/OpenModelicaLibraryTestingWork/OpenModelica/OMCompiler/build/bin/../lib/x86_64-linux-gnu/omc/libOpenModelicaRuntimeC.so(mmc_do_stackoverflow)
[bt] #3 /var/lib/jenkins/ws/OpenModelicaLibraryTestingWork/OpenModelica/OMCompiler/build/bin/../lib/x86_64-linux-gnu/omc/libOpenModelicaRuntimeC.so(GC.set.max.heap.size)
[bt] #4...5 /var/lib/jenkins/ws/OpenModelicaLibraryTestingWork/OpenModelica/OMCompiler/build/bin/../lib/x86_64-linux-gnu/omc/libOpenModelicaCompiler.so(NFModifier.Modifier.create)
[bt] #6 /var/lib/jenkins/ws/OpenModelicaLibraryTestingWork/OpenModelica/OMCompiler/build/bin/../lib/x86_64-linux-gnu/omc/libOpenModelicaCompiler.so(NFModifier.Modifier.fromElement)
[bt] #7 /var/lib/jenkins/ws/OpenModelicaLibraryTestingWork/OpenModelica/OMCompiler/build/bin/../lib/x86_64-linux-gnu/omc/libOpenModelicaCompiler.so(NFInst.instElementModifier)
[bt] #8 /var/lib/jenkins/ws/OpenModelicaLibraryTestingWork/OpenModelica/OMCompiler/build/bin/../lib/x86_64-linux-gnu/omc/libOpenModelicaCompiler.so(NFInst.instClassDef)
[bt] #9 /var/lib/jenkins/ws/OpenModelicaLibraryTestingWork/OpenModelica/OMCompiler/build/bin/../lib/x86_64-linux-gnu/omc/libOpenModelicaCompiler.so(NFInst.instClass)
```
@perost I'm not sure if this is an easy fix, but it would be worth having a look. Most other issues with the IDEAS library are due to not compliant Modelica code and will be fixed soon by @Mathadon, so I think it's important that we fix any remaining NF issues.

**Comment by Mathadon** [comment]
@casella it would be interesting to check further what the scaling limits are of OM. A good goal would be to support ~100 zones: 5000 state variables and 40k algebraic variables.

**Comment by perost** [comment]
The model works fine on my machine and uses about 11 GB of RAM to flatten, so there's at least no other issue than using a lot of memory and presumably hitting the limit we set in the library testing environment. Unfortunately it's really hard to reason about memory usage in MetaModelica, but I will see if I can maybe figure out where all the memory is going.

**Comment by Mathadon** [comment]
@perost this model uses the `RectangularZoneTemplate`, which includes many conditional components. If they are parsed then that could be part of the explanation. I've also seen a significant reduction in performance in Dymola for that model once we added support for interzonal airflow. That feature required that our `PropsBus` connectors include a Medium declaration. A Medium is a package and consequently each of those connectors carries quite a bunch of overhead. This may be (one of) the cause(s).

**Comment by casella** [comment]
11 GB these days is not such a big deal. For testing, though, we need to limit the memory footprint because we are running many simulation in parallel and we must be careful not to run out of memory.

From my experience 40k variables is no big deal, while 5000 states could be the problem.

@Mathadon you can turn on the -d=execstat compiler flag and see the memory allocation of each step in code generation.

**Comment by Mathadon** [comment]
@casella are those states a problem for the compiler or for the time integrator? We tend to use Euler integration to cope with the large number of state variables so if there is no bottleneck in the compiler then it sounds like we're ok since Euler scales linearly.

**Comment by casella** [comment]
Currently it is a problem with the compiler, but it could also be for the runtime simulator. 

I'm not sure if those systems are stiff or not. For sure DASSL with dense solver is a no-no for such large number of states: just holding the dense Jacobian would need about 200 MB, and inverting it would take forever. You can try the `ida` solver, that uses the sparse Kinsol/KLU solvers to solve the BDF formula. In our experience with the [ScalableTestSuite](https://libraries.openmodelica.org/branches/newInst-daemode/ScalableTestSuite_noopt/), that also scales linearly in most cases, so it's not bad - if it helps increasing the step size, it could eventually be faster.

BTW, @bernhardbachmann is developing a general-purpose adaptive step-size solver, as soon as that is merged in, you could experiment with something a bit better than explicit Euler.

**Comment by casella** [comment]
> @perost this model uses the `RectangularZoneTemplate`, which includes many conditional components. If they are flattened and not simply ignored then that could be part of the explanation. I've also seen a significant reduction in performance in Dymola for the RectangularZoneTemplate once we added support for interzonal airflow. That feature required that our `PropsBus` connectors include a Medium declaration. A Medium is a package and consequently each of those connectors carries quite a bunch of overhead.

As I understand, this shouldn't be a problem for the new frontend, which has been optimized to avoid such kind of overhead. @perost, can you comment on that.

>I will see if I can maybe figure out where all the memory is going

Yeah, it would be nice to have some clue if there is some bottleneck that can be removed with a more efficient implementation.

**Comment by perost** [comment]
> > @perost this model uses the `RectangularZoneTemplate`, which includes many conditional components. If they are flattened and not simply ignored then that could be part of the explanation. I've also seen a significant reduction in performance in Dymola for the RectangularZoneTemplate once we added support for interzonal airflow. That feature required that our `PropsBus` connectors include a Medium declaration. A Medium is a package and consequently each of those connectors carries quite a bunch of overhead.
> 
> As I understand, this shouldn't be a problem for the new frontend, which has been optimized to avoid such kind of overhead. @perost, can you comment on that.

We instantiate conditional components and remove them during the flattening phase, since that's the first place where we can do it. So if there are a lot of conditional components that are deleted that could be a problem. I had a quick look and out of about 50000 components that had a condition only about 2000 were deleted, but that doesn't say anything about the size of the components.

Ideally we'd delay instantiation of conditional components until we can evaluate their condition, but that might be complicated. I will experiment a bit and see.

I also played around a bit with changing how we store attributes of builtin types (`start`, `min`, `nominal`, etc), because right now we store all the attributes for every instance in order to allow modification of them. This seems to account for about half the memory usage. I've been wanting to redesign this for a long time so that we only store the attributes that actually have bindings.

**Comment by perost** [comment]
I had a closer look at the conditional components, and if I set the compiler to ignore the conditions the flat model becomes 5 times larger. So that's quite a lot more than I imagined and obviously something we need to fix.

I've started to change the way we handle conditional components so that we delay instantiating them until we can evaluate the conditions. Flattening the Heating model from IDEAS already works if I ignore connections and disable some connector things, this uses about 2 GB and takes a quarter of the earlier time. This is as I said without connect equations, but those usually don't take much time and memory.

The issue with connections is that they may reference conditional components, so we have to also delay instantiation of those until we've dealt with the conditional components (otherwise we can't e.g. look up connectors inside not deleted components).

Another issue is that we build some structures for connectors and e.g. check that they are balanced, which also needs to be delayed if we have conditional components in the connectors (which the Heating model has).

So I think this will work and greatly improve these models and possibly others too, I don't really know if other libraries use conditional components as extensively. But as mentioned there are some issues left for me to solve, so it might be a while (particularly since tomorrow is also a Swedish holiday).

**Comment by Mathadon** [comment]
@perost to clarify, the RectangularZoneTemplate is something we do not intend to use much in the future and other models don't have as many conditions components. So for IDEAS I wouldn't consider this as high priority but improvements are always welcome of course.

**Comment by casella** [comment]
@perost we are planning to build a library to support boiler design, to be used with OMC. The heat exchangers models in that library can be either 1D or 2D. The 1D model is small, the 2D model contains _a lot_ of equations. The way we'd like to design the library is to have both as conditionally activated sub-models - note that if we use the 2D model, we also need the 1D model to get good start values for the 2D model successful initialization.

The idea is that most of the time, most of the components will just use the 1D version, with 2D being turned on occasionally for some components only. Hence, for this application, avoiding the unnecessary instantiation of disabled components will be absolutely essential to avoid long code generation times.

This is by no way urgent, but if this project goes ahead, it will become very important. So, I would definitely spend some time on it, but maybe not with top priority.

It would be good if you opened a separate ticket describing this enhancement, which has a far greater scope than the IDEAS out-of-memory issue.

**Comment by casella** [comment]
See #9038.
